core/protobuf/src/main/java/alluxio/util/proto/ProtoMessage.java

```diff
@@ -23,73 +23,67 @@
  * different generated messages. This class is intended to be used internally only.
  */
 public final class ProtoMessage {
-
-  /**
-   * Type of the message.
-   */
-  public enum Type {
-    READ_REQUEST,
-    WRITE_REQUEST,
-    RESPONSE,
-  }
-
   private MessageLite mMessage;
-  private Type mType;
 
   /**
-   * Constructs a {@link ProtoMessage} instance wrapping around {@link Protocol.ReadRequest}.
+   * Constructs a {@link ProtoMessage} instance wrapping around {@link MessageLite}.
    *
    * @param message the message to wrap
    */
-  public ProtoMessage(Protocol.ReadRequest message) {
-    this(message, Type.READ_REQUEST);
+  public ProtoMessage(MessageLite message) {
+    mMessage = message;
   }
 
   /**
-   * Constructs a {@link ProtoMessage} instance wrapping around {@link Protocol.WriteRequest}.
+   * Gets the protobuf message given a type. Returns null if the mMessage is not of type tClass.
+   * This is slow since it uses reflection.
    *
-   * @param message the message to wrap
-   */
-  public ProtoMessage(Protocol.WriteRequest message) {
-    this(message, Type.WRITE_REQUEST);
-  }
-
-  /**
-   * Constructs a {@link ProtoMessage} instance wrapping around {@link Protocol.Response}.
+   * @param <T> the type T
    *
-   * @param message the message to wrap
+   * @return the unwrapped message as type T
    */
-  public ProtoMessage(Protocol.Response message) {
-    this(message, Type.RESPONSE);
+  <T> T getMessage(Class<T> tClass) {
+    if (tClass.isAssignableFrom(mMessage.getClass())) {
+      return (T) mMessage;
+    } else {
+      return null;
+    }
   }
 
   /**
-   * Constructs a {@link ProtoMessage} instance wrapping around {@link MessageLite}.
+   * Gets the read request or null if mMessage is not of type {@link Protocol.ReadRequest}.
    *
-   * @param message the message to wrap
-   * @param type type of the message
+   * @return the read request or null
    */
-  public ProtoMessage(MessageLite message, Type type) {
-    mMessage = message;
-    mType = type;
+  public Protocol.ReadRequest asReadRequest() {
```

Comment:
suggestion: I would have two APIs related to each proto type:
`IsFoo()` which returns whether this proto message is of type
and `AsFoo` which returns the proto instance in the type, or throw runtime error